### DISCUSSION TITLE
503 on default server root path

### DISCUSSION THREAD

**Discussion by markfaine** [nginx-proxy/nginx-proxy#2122] [Q&A]
## Questions

If you have a question, DO NOT SUBMIT a new issue.  Please ask the question on the Q&A Group: https://groups.google.com/forum/#!forum/nginx-proxy

Group does not exist

![image](https://user-images.githubusercontent.com/949949/91491414-9b280700-e879-11ea-98da-2d53b1e9ff03.png)

I am trying to get the default path https://host.domain.com/ to serve the content in /usr/share/nginx/html.  This is not a vhost but the main server itself.  I was able to get it to start with https by renaming the ssl certs as "default" but it always returns 503 Service Temporarily Unavailable

What is required to get it to serve this content?  I do of course plan on adding virtual hosts but I need the default root to also serve content on 443. 

Thanks

**Comment by markfaine** [comment]
Update: I have been able to workaround it:

I modified default.conf to remove the 503 return

```
server {
        server_name _; # This is just an invalid value which will never trigger on a real hostname.
        listen 80;
        access_log /var/log/nginx/access.log vhost;
        # Do not HTTPS redirect Let'sEncrypt ACME challenge
        location /.well-known/acme-challenge/ {
                auth_basic off;
                allow all;
                root /usr/share/nginx/html;
                try_files $uri =404;
                break;
        }
        location / {
                return 301 https://$host$request_uri;
        }

}
server {
        server_name _; # This is just an invalid value which will never trigger on a real hostname.
        listen 443 ssl http2;
        access_log /var/log/nginx/access.log vhost;
        root /usr/share/nginx/html;
        ssl_session_cache shared:SSL:50m;
        ssl_session_tickets off;
        ssl_certificate /etc/nginx/certs/default.crt;
        ssl_certificate_key /etc/nginx/certs/default.key;
}

```

How do I ensure that this part of default.conf isn't changed.  Should I update nginx.tmpl or is there a builtin way to do it?

Thanks!

**Comment by tkw1536** [comment]
From what I understand you want to serve static content on the domain belonging to the server itself. Let's call that domain `host.domain.com`. 

There is nothing that makes this domain special, and you can serve it like you would any other. 
Therefore I don't think you have to modify `nginx.tmpl` to achieve this effect. 

I see two approaches you could take. 
For either approach, you should also rename your certificates to `host.domain.com.crt` and  `host.domain.com.key`. 

1. Create a docker container that has `VIRTUAL_HOST=host.domain.com` set and serves only the static files. There are many of those available, one being https://github.com/PierreZ/goStatic. This would be something like:

```bash
docker run -d -v /path/to/sources:/srv/http -e VIRTUAL_HOST=host.domain.com -e VIRTUAL_PORT=8043 pierrezemb/gostatic
```

2. You could make use of [Per-VIRTUAL_HOST location configuration](https://github.com/nginx-proxy/nginx-proxy#per-virtual_host-location-configuration). To achieve that you would create a dummy container, that doesn't serve anything but has the `VIRTUAL_HOST=host.domain.com` variable set. You could mount the file `/etc/nginx/vhost.d/host.domain.com_location` inside the nginx-proxy container and insert any directives that you want inside it (including serving static content). 

I would recommend (1), as it is much simpler to configure. 
Furthermore, you might also want to set it as the [`DEFAULT_HOST`](https://github.com/nginx-proxy/nginx-proxy#default-host), which causes any unknown domain (like e.g. connecting directly to the IP Address) to be treated as this domain.

**Comment by markfaine** [comment]
Thanks, so it's not possible to do without adding a container? Is this something that you would consider for a feature request?

-------- Original Message --------

**Comment by sgalcheung** [comment]
Maybe my front end is not deployed here, but when I visit the ip, it also returns 503 Service Temporarily Unavailable. I just don’t want him to return an error. Maybe it can map an index.html file like nginx.

**Comment by sgalcheung** [comment]
Now I closed port 80 of the server's firewall

and, I’m thinking about whether port 80 can be undisclosed in nginx-proxy

**Comment by adamu** [comment]
You can put your two server definitions in a config file and mount it to `/etc/nginx/conf.d/` on the nginx-proxy container.

If you replace the `server_name _` with `server_name host.domain.com`, it should take priorty over the `_` definition in `default.conf`. Another way might be to specify [`default_server`](https://nginx.org/en/docs/http/ngx_http_core_module.html#listen) (but I didn't try that, because I wanted to keep the default 503 behaviour for unknown hosts).

You might also need to make sure the config filename is alphabetically later than `default.conf`, by e.g. prefixing the filename with `z`, to make sure it loads after the default config (so that the `log_format vhost` definition is available, and so your config doesn't override the defaults). I'm not sure if this is defined behaviour, but it was necessary for me.

**Comment by davidparks21** [comment]
I have the same problem being discussed here, but I'm confused because the letsencrypt companion needs to serve static content for the `.well-known/acme-challenge/` files to get SSL certs generated automatically. That process is not working because the static content isn't being served by the nginx container (503 errors), so either I'm doing something very wrong, or the containers aren't configured to work together out of the box. I feel like I must be doing something wrong because I'm adding custom configurations like the ones discussed here just to get nginx to serve these static files on the root domain. I don't have a VIRTUAL_HOST defined on the root domain, all the services run on a sub domain.
